### PR TITLE
Feature fpr algorithm performance

### PR DESCRIPTION
**Pull request by coreyostrove**
This branch introduces a new greedy search algorithm for performing per-germ FPR. This algorithm leverages the same low-rank update machinery used to generate speedups for germ selection. Due to the nature of the low-rank updates used, this utilizes an objective function based on the trace of the psuedo-inverse of a candidate fiducial set's Jacobian instead of the minimum eigenvalue (these should largely track with each other though). In the course of developing this a number of edge cases were revealed in the low-rank update code which didn't arise when testing germ selection. These have been patched, and as a nice side-effect the fixes actually result in a modest performance boost for germ selection too.

Smaller changes included in the branch include:

- bugfixes for numpy data type handling
- bugfixes for leaky RNG seed plumbing
- implementation of a simple trick for SVDs of gramians that speeds up the construction of EVD caches
- Mass expunging of debugging and print statements that are no longer needed

### CODE REVIEW

**Review by sserita — COMMENTED** [review]
It looks good to me, although I haven't checked the math in detail. As Erik suggested, we can leave this open until we figure out the other FPR issue.

**Comment by coreyostrove** [comment]
I've spent some time looking into @enielse's concerns about the size of the fiducial sets returned for the randomized fiducial pair search routine. I've more or less convinced myself that this isn't a bug. I added additional logging to the randomized search routine so that the maximum rank seen for any Jacobian at each iteration (iteration = candidate fiducial set size) was printed. It looks like (for the particular test case using XYCNOT) that selecting fiducials at random just does a pretty poor job at giving us a full-rank Jacobian (this wouldn't have been my intuition, but it looks to be the case here). The objective function here can't be satisfied until the Jacobian is full-rank and so this is the limiting factor.

Looking into this did help me identify a mistake in the greedy algorithm, by the way. So this was all worthwhile. I had mistakenly failed to include in the exit criterion a check to ensure that the Jacobian was full rank and, unlike the minimum eigenvalue criterion, the psuedoinverse trace condition can be satisfied without being full rank because of how the psuedoinverse is defined. That is fixed now. We now get fiducial sets that are a bit larger, but the greedy search really does seem to be outperforming the random search by a significant amount here. The greedy score function does favor updates that increase the rank of the fiducial set the most up until the point where the Jacobian becomes full-rank, so maybe this isn't too surprising.

**Comment by enielse** [comment]
I did some digging too - and agree with @coreyostrove: I don't think there's necessarily a bug in the current FPR algorithm.  The large number of pairs the old method is finding differs from earlier pyGSTi versions but from what I can tell this is due to bug *fixes* rather than introductions.  Down the line, we should be able to trim down the list of needed pairs by tracking the parameter directions a germ was intended to amplify, but that's for another PR.

**Review by sserita — APPROVED** [review]
All looks good to me!

File: pygsti/algorithms/germselection.py [hunk 1]
```diff
@@ -3101,98 +3103,119 @@ def minamide_style_inverse_trace(update, orig_e, U, proj_U, force_rank_increase=
...
+        #Identity matrix in the central matrix we're inverting should have dimension
+        #equal to the number of columns in beta (or equivalently the update matrix)
+        central_mat= _np.linalg.inv(_np.eye(beta.shape[1]) + pinv_sqrt_E_beta.T@pinv_sqrt_E_beta)
+        
+        #diagnostic information
+        if central_mat.shape == (0,0):
+            print('central_mat shape: ', central_mat.shape)
+        #now calculate the diagonal elements of pinv_E_beta@central_mat@pinv_E_beta.T
+        
+        #Take the cholesky decomposition of the central matrix:
```
[the remark this answers is not in the document]
Comment (coreyostrove) [reply pygsti/algorithms/germselection.py]: Thanks for the feedback, @enielse. I added some comments on this in my latest commit.

**Comment by coreyostrove** [comment]
I finally forced myself to sit down and write up a description of the theory behind the low-rank update code. I eventually plan to convert this into an expanded and more detailed write-up in latex (hence why it reads the way it does), so let me know if anything is unclear or confusing and I'll try to clarify that once I get around to writing that document.

**Pull request merged**